### ISSUE TITLE
Shader compilation failure on tvOS

### ISSUE THREAD

**Issue by warmenhoven** [KhronosGroup/MoltenVK#2329]
Starting with c37a880bebad1286113dc1ab9e412080136f3046, on tvOS I'm getting a shader compilation failure for the following shader:

```
#version 450

layout(local_size_x = 16, local_size_y = 8) in;

layout(set = 0, binding = 3, rgba8) writeonly uniform image2D uImage;
layout(set = 0, binding = 4, std430) readonly buffer SSBO
{
	uint packed_pixels[];
};

layout(set = 0, binding = 0, std140) uniform UBO
{
	uvec2 resolution;
	uint word_stride;
};

vec3 rgb565_to_rgb888(uint word)
{
	return vec3((uvec3(word) >> uvec3(11, 5, 0)) & uvec3(31, 63, 31)) / vec3(31.0, 63.0, 31.0);
}

void main()
{
	// We work on two horizonal pixels in parallel since we cannot rely on 16-bit storage.
	uvec2 first_input_pixel = gl_GlobalInvocationID.xy;
	uvec2 first_output_pixel = first_input_pixel * uvec2(2, 1);

	if (all(lessThan(first_output_pixel, resolution)))
	{
		uint word = packed_pixels[first_input_pixel.y * word_stride + first_input_pixel.x];
		uint lo_word = word & 0xffffu;
		uint hi_word = word >> 16u;
		vec3 lo = rgb565_to_rgb888(lo_word);
		vec3 hi = rgb565_to_rgb888(hi_word);

		imageStore(uImage, ivec2(first_output_pixel), vec4(lo, 1.0));
		if (first_output_pixel.x + 1u < resolution.x)
			imageStore(uImage, ivec2(first_output_pixel) + ivec2(1, 0), vec4(hi, 1.0));
	}
}
```

Here's a log of the conversion and compilation:

```
[mvk-info] Converting SPIR-V:
; SPIR-V
; Version: 1.0
; Generator: Google Shaderc over Glslang; 11
; Bound: 129
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %4 "main" %37
               OpExecutionMode %4 LocalSize 16 8 1
               OpSource GLSL 450
               OpSourceExtension "GL_GOOGLE_cpp_style_line_directive"
               OpSourceExtension "GL_GOOGLE_include_directive"
               OpName %4 "main"
               OpName %12 "rgb565_to_rgb888(u1;"
               OpName %11 "word"
               OpName %35 "first_input_pixel"
               OpName %37 "gl_GlobalInvocationID"
               OpName %40 "first_output_pixel"
               OpName %47 "UBO"
               OpMemberName %47 0 "resolution"
               OpMemberName %47 1 "word_stride"
               OpName %49 ""
               OpName %61 "word"
               OpName %63 "SSBO"
               OpMemberName %63 0 "packed_pixels"
               OpName %65 ""
               OpName %78 "lo_word"
               OpName %82 "hi_word"
               OpName %87 "lo"
               OpName %88 "param"
               OpName %91 "hi"
               OpName %92 "param"
               OpName %97 "uImage"
               OpDecorate %37 BuiltIn GlobalInvocationId
               OpMemberDecorate %47 0 Offset 0
               OpMemberDecorate %47 1 Offset 8
               OpDecorate %47 Block
               OpDecorate %49 DescriptorSet 0
               OpDecorate %49 Binding 0
               OpDecorate %62 ArrayStride 4
               OpMemberDecorate %63 0 NonWritable
               OpMemberDecorate %63 0 Offset 0
               OpDecorate %63 BufferBlock
               OpDecorate %65 DescriptorSet 0
               OpDecorate %65 Binding 4
               OpDecorate %97 DescriptorSet 0
               OpDecorate %97 Binding 3
               OpDecorate %97 NonReadable
               OpDecorate %128 BuiltIn WorkgroupSize
          %2 = OpTypeVoid
          %3 = OpTypeFunction %2
          %6 = OpTypeInt 32 0
          %7 = OpTypePointer Function %6
          %8 = OpTypeFloat 32
          %9 = OpTypeVector %8 3
         %10 = OpTypeFunction %9 %7
         %15 = OpTypeVector %6 3
         %17 = OpConstant %6 11
         %18 = OpConstant %6 5
         %19 = OpConstant %6 0
         %20 = OpConstantComposite %15 %17 %18 %19
         %22 = OpConstant %6 31
         %23 = OpConstant %6 63
         %24 = OpConstantComposite %15 %22 %23 %22
         %27 = OpConstant %8 31
         %28 = OpConstant %8 63
         %29 = OpConstantComposite %9 %27 %28 %27
         %33 = OpTypeVector %6 2
         %34 = OpTypePointer Function %33
         %36 = OpTypePointer Input %15
         %37 = OpVariable %36 Input
         %42 = OpConstant %6 2
         %43 = OpConstant %6 1
         %44 = OpConstantComposite %33 %42 %43
         %47 = OpTypeStruct %33 %6
         %48 = OpTypePointer Uniform %47
         %49 = OpVariable %48 Uniform
         %50 = OpTypeInt 32 1
         %51 = OpConstant %50 0
         %52 = OpTypePointer Uniform %33
         %55 = OpTypeBool
         %56 = OpTypeVector %55 2
         %62 = OpTypeRuntimeArray %6
         %63 = OpTypeStruct %62
         %64 = OpTypePointer Uniform %63
         %65 = OpVariable %64 Uniform
         %68 = OpConstant %50 1
         %69 = OpTypePointer Uniform %6
         %80 = OpConstant %6 65535
         %84 = OpConstant %6 16
         %86 = OpTypePointer Function %9
         %95 = OpTypeImage %8 2D 0 0 0 2 Rgba8
         %96 = OpTypePointer UniformConstant %95
         %97 = OpVariable %96 UniformConstant
        %100 = OpTypeVector %50 2
        %103 = OpConstant %8 1
        %104 = OpTypeVector %8 4
        %120 = OpConstantComposite %100 %68 %51
        %127 = OpConstant %6 8
        %128 = OpConstantComposite %15 %84 %127 %43
          %4 = OpFunction %2 None %3
          %5 = OpLabel
         %35 = OpVariable %34 Function
         %40 = OpVariable %34 Function
         %61 = OpVariable %7 Function
         %78 = OpVariable %7 Function
         %82 = OpVariable %7 Function
         %87 = OpVariable %86 Function
         %88 = OpVariable %7 Function
         %91 = OpVariable %86 Function
         %92 = OpVariable %7 Function
         %38 = OpLoad %15 %37
         %39 = OpVectorShuffle %33 %38 %38 0 1
               OpStore %35 %39
         %41 = OpLoad %33 %35
         %45 = OpIMul %33 %41 %44
               OpStore %40 %45
         %46 = OpLoad %33 %40
         %53 = OpAccessChain %52 %49 %51
         %54 = OpLoad %33 %53
         %57 = OpULessThan %56 %46 %54
         %58 = OpAll %55 %57
               OpSelectionMerge %60 None
               OpBranchConditional %58 %59 %60
         %59 = OpLabel
         %66 = OpAccessChain %7 %35 %43
         %67 = OpLoad %6 %66
         %70 = OpAccessChain %69 %49 %68
         %71 = OpLoad %6 %70
         %72 = OpIMul %6 %67 %71
         %73 = OpAccessChain %7 %35 %19
         %74 = OpLoad %6 %73
         %75 = OpIAdd %6 %72 %74
         %76 = OpAccessChain %69 %65 %51 %75
         %77 = OpLoad %6 %76
               OpStore %61 %77
         %79 = OpLoad %6 %61
         %81 = OpBitwiseAnd %6 %79 %80
               OpStore %78 %81
         %83 = OpLoad %6 %61
         %85 = OpShiftRightLogical %6 %83 %84
               OpStore %82 %85
         %89 = OpLoad %6 %78
               OpStore %88 %89
         %90 = OpFunctionCall %9 %12 %88
               OpStore %87 %90
         %93 = OpLoad %6 %82
               OpStore %92 %93
         %94 = OpFunctionCall %9 %12 %92
               OpStore %91 %94
         %98 = OpLoad %95 %97
         %99 = OpLoad %33 %40
        %101 = OpBitcast %100 %99
        %102 = OpLoad %9 %87
        %105 = OpCompositeExtract %8 %102 0
        %106 = OpCompositeExtract %8 %102 1
        %107 = OpCompositeExtract %8 %102 2
        %108 = OpCompositeConstruct %104 %105 %106 %107 %103
               OpImageWrite %98 %101 %108
        %109 = OpAccessChain %7 %40 %19
        %110 = OpLoad %6 %109
        %111 = OpIAdd %6 %110 %43
        %112 = OpAccessChain %69 %49 %51 %19
        %113 = OpLoad %6 %112
        %114 = OpULessThan %55 %111 %113
               OpSelectionMerge %116 None
               OpBranchConditional %114 %115 %116
        %115 = OpLabel
        %117 = OpLoad %95 %97
        %118 = OpLoad %33 %40
        %119 = OpBitcast %100 %118
        %121 = OpIAdd %100 %119 %120
        %122 = OpLoad %9 %91
        %123 = OpCompositeExtract %8 %122 0
        %124 = OpCompositeExtract %8 %122 1
        %125 = OpCompositeExtract %8 %122 2
        %126 = OpCompositeConstruct %104 %123 %124 %125 %103
               OpImageWrite %117 %121 %126
               OpBranch %116
        %116 = OpLabel
               OpBranch %60
         %60 = OpLabel
               OpReturn
               OpFunctionEnd
         %12 = OpFunction %9 None %10
         %11 = OpFunctionParameter %7
         %13 = OpLabel
         %14 = OpLoad %6 %11
         %16 = OpCompositeConstruct %15 %14 %14 %14
         %21 = OpShiftRightLogical %15 %16 %20
         %25 = OpBitwiseAnd %15 %21 %24
         %26 = OpConvertUToF %9 %25
         %30 = OpFDiv %9 %26 %29
               OpReturnValue %30
               OpFunctionEnd

End SPIR-V

Converted MSL:
#pragma clang diagnostic ignored "-Wmissing-prototypes"

#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct UBO
{
    uint2 resolution;
    uint word_stride;
};

struct SSBO
{
    uint packed_pixels[1];
};

constant uint3 gl_WorkGroupSize [[maybe_unused]] = uint3(16u, 8u, 1u);

static inline __attribute__((always_inline))
float3 rgb565_to_rgb888(thread const uint& word)
{
    return float3((uint3(word) >> uint3(11u, 5u, 0u)) & uint3(31u, 63u, 31u)) / float3(31.0, 63.0, 31.0);
}

kernel void main0(constant UBO& _49 [[buffer(9)]], const device SSBO& _65 [[buffer(9)]], texture2d<float, access::write> uImage [[texture(0)]], uint3 gl_GlobalInvocationID [[thread_position_in_grid]])
{
    uint2 first_input_pixel = gl_GlobalInvocationID.xy;
    uint2 first_output_pixel = first_input_pixel * uint2(2u, 1u);
    if (all(first_output_pixel < _49.resolution))
    {
        uint word = _65.packed_pixels[(first_input_pixel.y * _49.word_stride) + first_input_pixel.x];
        uint lo_word = word & 65535u;
        uint hi_word = word >> 16u;
        uint param = lo_word;
        float3 lo = rgb565_to_rgb888(param);
        uint param_1 = hi_word;
        float3 hi = rgb565_to_rgb888(param_1);
        uImage.write(float4(lo, 1.0), uint2(int2(first_output_pixel)));
        if ((first_output_pixel.x + 1u) < _49.resolution.x)
        {
            uImage.write(float4(hi, 1.0), uint2((int2(first_output_pixel) + int2(1, 0))));
        }
    }
}


End MSL

Estimated original GLSL:
#version 450
layout(local_size_x = 16, local_size_y = 8, local_size_z = 1) in;

layout(set = 0, binding = 0, std140) uniform UBO
{
    uvec2 resolution;
    uint word_stride;
} _49;

layout(set = 0, binding = 4, std430) readonly buffer SSBO
{
    uint packed_pixels[];
} _65;

layout(set = 0, binding = 3, rgba8) uniform writeonly image2D uImage;

vec3 rgb565_to_rgb888(uint word)
{
    return vec3((uvec3(word) >> uvec3(11u, 5u, 0u)) & uvec3(31u, 63u, 31u)) / vec3(31.0, 63.0, 31.0);
}

void main()
{
    uvec2 first_input_pixel = gl_GlobalInvocationID.xy;
    uvec2 first_output_pixel = first_input_pixel * uvec2(2u, 1u);
    if (all(lessThan(first_output_pixel, _49.resolution)))
    {
        uint word = _65.packed_pixels[(first_input_pixel.y * _49.word_stride) + first_input_pixel.x];
        uint lo_word = word & 65535u;
        uint hi_word = word >> 16u;
        uint param = lo_word;
        vec3 lo = rgb565_to_rgb888(param);
        uint param_1 = hi_word;
        vec3 hi = rgb565_to_rgb888(param_1);
        imageStore(uImage, ivec2(first_output_pixel), vec4(lo, 1.0));
        if ((first_output_pixel.x + 1u) < _49.resolution.x)
        {
            imageStore(uImage, ivec2(first_output_pixel) + ivec2(1, 0), vec4(hi, 1.0));
        }
    }
}


End GLSL


[mvk-info] Compiling Metal shader with FastMath enabled.
Compiler failed to build request
[mvk-error] VK_ERROR_INITIALIZATION_FAILED: Shader library compile failed (Error code 3):
program_source:27:71: error: cannot reserve 'buffer' resource location at index 9
kernel void main0(constant UBO& _49 [[buffer(9)]], const device SSBO& _65 [[buffer(9)]], texture2d<float, access::write> uImage [[texture(0)]], uint3 gl_GlobalInvocationID [[thread_position_in_grid]])
                                                                      ^
.
[mvk-error] VK_ERROR_INVALID_SHADER_NV: Compute shader function could not be compiled into pipeline. See previous logged error.
```

Please let me know if there's any additional information I can provide.

**Comment by warmenhoven** [comment]
Setting MVK_CONFIG_USE_METAL_ARGUMENT_BUFFERS back to 0 works around this for me.

**Comment by billhollings** [comment]
PR #2335 fixes this issue. Please test with latest MoltenVK, and close this issue if the problem is resolved.

**Comment by warmenhoven** [comment]
Thank you! That PR fixes all of my problems.

**Comment by warmenhoven** [comment]
Not all of my problems it seems; it's now leaking about 22MB of GPU memory per second, and only on the old hardware I have. I'll open a new issue about it.
